### PR TITLE
LVM Optimization: Relation Encoding

### PR DESCRIPTION
**Pull request by XiaowenHu96**
With this PR, LVM will encode relation into an index position, instead of using a mapping from string to relation pointer.

This speeds up the LVM.

### CODE REVIEW

File: src/LVMGenerator.h [hunk 1]
```diff
@@ -21,6 +21,41 @@
...
+        return relationNames[relId];
+    }
+
+    /** Get total number of relations */
+    size_t getSize() const {
+        return relationNames.size();
+    }
+
+    /** RelName to index mapping */
+    std::map<std::string, size_t> relations;
```
Comment (mmcgr) [inline src/LVMGenerator.h]: Are the variables intended to be public? If not, make this a `class`

File: src/LVM.cpp [hunk 2]
```diff
@@ -897,38 +903,38 @@ void LVM::execute(std::unique_ptr<LVMCode>& codeStream, InterpreterContext& ctxt
...
+                size_t sourceId = code[ip + 1];
+                size_t targetId = code[ip + 2];
                 // get involved relation
-                InterpreterRelation& src = getRelation(source);
-                InterpreterRelation& trg = getRelation(target);
+                auto srcPtr = getRelation(sourceId);
+                auto trgPtr = getRelation(targetId);
 
-                if (dynamic_cast<InterpreterEqRelation*>(&trg)) {
+                if (dynamic_cast<InterpreterEqRelation*>(trgPtr)) {
```
Comment (mmcgr) [inline src/LVM.cpp]: Explicitly compare to nullptr
```suggestion
                if (dynamic_cast<InterpreterEqRelation*>(trgPtr) != nullptr) {
```

**Comment by b-scholz** [comment]
@XiaowenHu96 - please can you resolve the conflict. Also please can you duplicate the InterpreterRelation/InterpreterIndex classes for LVM so that we don't need to share a common interface with RAMI anymore (it will change very soon quickly). 

I would suggest that we rename the files 

``InterpreterContext.h    InterpreterInterface.h  InterpreterRelation.h
Interpreter.h           InterpreterRecords.cpp  
InterpreterIndex.h      InterpreterRecords.h ``

to ```RAMI<name>``` (instead of ```Interpreter<name>```). 

Duplicate the same files for LVM and call them ```LVM<name>```

I feel that we reached a point were we need to separate both interpreters.

**Comment by XiaowenHu96** [comment]
@b-scholz Yes. I'll make another PR for this.

**Pull request merged**